File: packages/types/src/event.ts
import Ajv, {ValidateFunction} from "ajv";
import eventSchema from "../../../schema/event.v1.json";
import agentFrameSchema from "../../../schema/agent_frame.v1.json";

export type MinimalParameters = {
  POINTER_DELTA: { dx: number; dy: number };
  ZOOM_DELTA: number;
  ROT_DELTA: number;
  INPUT_TRIGGER: boolean;
  HOLO_FRAME?: HoloFrame;
};

export type HoloFrame = {
  quaternion: [number, number, number, number];
  translation: [number, number, number];
  surface: string;
  metadata?: Record<string, unknown>;
};

export type EventEnvelope = {
  type: "input" | "holo_frame" | "agent";
  timestamp: number;
  payload: MinimalParameters;
};

export type AgentFrame = {
  role: string;
  goal: string;
  sdk_surface: string;
  bounds: { x: number; y: number; z: number };
  focus: { path: string; hoist?: boolean };
  inputs: MinimalParameters;
  outputs: string[];
  safety: { spawn_bounds: number; rate_limit: number; rejection_reason?: string };
  telemetry?: { latency_ms?: number; jitter_ms?: number; replay_frame?: string };
};

export type EnvelopeKind = "event.v1" | "agent_frame.v1";

const ajv = new Ajv({allErrors: true, strict: false});
const validateEvent = ajv.compile<EventEnvelope>(eventSchema as any) as ValidateFunction<EventEnvelope>;
const validateAgentFrame = ajv.compile<AgentFrame>(agentFrameSchema as any) as ValidateFunction<AgentFrame>;

export function validator(kind: EnvelopeKind): ValidateFunction<EventEnvelope | AgentFrame> {
  if (kind === "event.v1") return validateEvent;
  if (kind === "agent_frame.v1") return validateAgentFrame;
  throw new Error(`Unknown envelope kind: ${kind}`);
}

export function assertValid(kind: EnvelopeKind, payload: EventEnvelope | AgentFrame): void {
  const validate = validator(kind);
  const valid = validate(payload);
  if (!valid) {
    const message = validate.errors?.map((err) => `${err.instancePath} ${err.message}`).join("; ") || "validation failed";
    throw new Error(message);
  }
}

File: packages/types_dart/lib/event.dart
class MinimalParameters {
  final double dx;
  final double dy;
  final double zoomDelta;
  final double rotDelta;
  final bool inputTrigger;
  final HoloFrame? holoFrame;

  const MinimalParameters({
    required this.dx,
    required this.dy,
    required this.zoomDelta,
    required this.rotDelta,
    required this.inputTrigger,
    this.holoFrame,
  });

  factory MinimalParameters.fromJson(Map<String, dynamic> json) {
    return MinimalParameters(
      dx: (json['POINTER_DELTA']['dx'] as num).toDouble(),
      dy: (json['POINTER_DELTA']['dy'] as num).toDouble(),
      zoomDelta: (json['ZOOM_DELTA'] as num).toDouble(),
      rotDelta: (json['ROT_DELTA'] as num).toDouble(),
      inputTrigger: json['INPUT_TRIGGER'] as bool,
      holoFrame: json['HOLO_FRAME'] != null ? HoloFrame.fromJson(json['HOLO_FRAME']) : null,
    );
  }

  Map<String, dynamic> toJson() => {
        'POINTER_DELTA': {'dx': dx, 'dy': dy},
        'ZOOM_DELTA': zoomDelta,
        'ROT_DELTA': rotDelta,
        'INPUT_TRIGGER': inputTrigger,
        if (holoFrame != null) 'HOLO_FRAME': holoFrame!.toJson(),
      };
}

class HoloFrame {
  final List<double> quaternion;
  final List<double> translation;
  final String surface;
  final Map<String, dynamic>? metadata;

  const HoloFrame({
    required this.quaternion,
    required this.translation,
    required this.surface,
    this.metadata,
  });

  factory HoloFrame.fromJson(Map<String, dynamic> json) {
    return HoloFrame(
      quaternion: (json['quaternion'] as List).map((e) => (e as num).toDouble()).toList(),
      translation: (json['translation'] as List).map((e) => (e as num).toDouble()).toList(),
      surface: json['surface'] as String,
      metadata: json['metadata'] != null ? Map<String, dynamic>.from(json['metadata'] as Map) : null,
    );
  }

  Map<String, dynamic> toJson() => {
        'quaternion': quaternion,
        'translation': translation,
        'surface': surface,
        if (metadata != null) 'metadata': metadata,
      };
}

class EventEnvelope {
  final String type;
  final double timestamp;
  final MinimalParameters payload;

  const EventEnvelope({required this.type, required this.timestamp, required this.payload});

  factory EventEnvelope.fromJson(Map<String, dynamic> json) {
    return EventEnvelope(
      type: json['type'] as String,
      timestamp: (json['timestamp'] as num).toDouble(),
      payload: MinimalParameters.fromJson(json['payload'] as Map<String, dynamic>),
    );
  }

  Map<String, dynamic> toJson() => {
        'type': type,
        'timestamp': timestamp,
        'payload': payload.toJson(),
      };
}

class AgentFrame {
  final String role;
  final String goal;
  final String sdkSurface;
  final Map<String, double> bounds;
  final Map<String, dynamic> focus;
  final MinimalParameters inputs;
  final List<String> outputs;
  final Map<String, dynamic> safety;
  final Map<String, dynamic>? telemetry;

  const AgentFrame({
    required this.role,
    required this.goal,
    required this.sdkSurface,
    required this.bounds,
    required this.focus,
    required this.inputs,
    required this.outputs,
    required this.safety,
    this.telemetry,
  });

  factory AgentFrame.fromJson(Map<String, dynamic> json) {
    return AgentFrame(
      role: json['role'] as String,
      goal: json['goal'] as String,
      sdkSurface: json['sdk_surface'] as String,
      bounds: (json['bounds'] as Map).map((k, v) => MapEntry(k.toString(), (v as num).toDouble())),
      focus: Map<String, dynamic>.from(json['focus'] as Map),
      inputs: MinimalParameters.fromJson(json['inputs'] as Map<String, dynamic>),
      outputs: (json['outputs'] as List).cast<String>(),
      safety: Map<String, dynamic>.from(json['safety'] as Map),
      telemetry: json['telemetry'] != null ? Map<String, dynamic>.from(json['telemetry'] as Map) : null,
    );
  }

  Map<String, dynamic> toJson() => {
        'role': role,
        'goal': goal,
        'sdk_surface': sdkSurface,
        'bounds': bounds,
        'focus': focus,
        'inputs': inputs.toJson(),
        'outputs': outputs,
        'safety': safety,
        if (telemetry != null) 'telemetry': telemetry,
      };
}
